src/Oscoin/Data/OscoinTx.hs

```diff
@@ -0,0 +1,331 @@
+    if WorldState.member key ws
+       then Left (ErrProjectExists addr)
+       else Right ( WorldState.insert key (ProjectVal (mkProject addr)) ws
+                  , TxOutput )
+  where
+    key = addressKey addr
+
+applyTxPayload (TxUnregisterProject addr) author ws = do
+    p@Project{..} <- lookupProject addr   ws
+    member        <- lookupMember  author ws
+
+    mapHandlerError (pUnregister p member)
+    pure (WorldState.delete projKey ws, TxOutput)
+  where
+    projKey = addressKey addr
+
+applyTxPayload (TxAuthorize addr pk) author ws = do
+    p@Project{..} <- lookupProject addr    ws
+    member        <- lookupMember  author  ws
+    epoch         <- lookupEpoch           ws
+
+    mapHandlerError (pAuthorize pk p member)
+    pure (adjust (addKey epoch) projKey ws, TxOutput)
+  where
+    projKey = addressKey addr
+
+    addKey e v = case v of
+        ProjectVal p -> ProjectVal $
+            p { pMaintainers = Map.insert pk (mkMember pk e) (pMaintainers p) }
+        _ -> v
+
+applyTxPayload (TxDeauthorize addr pk) author ws = do
+    p@Project{..} <- lookupProject addr    ws
+    member        <- lookupMember  author  ws
+
+    mapHandlerError (pDeauthorize pk p member)
+    pure (adjust removeKey projKey ws, TxOutput)
+  where
+    projKey = addressKey addr
+
+    removeKey v = case v of
+        ProjectVal p -> ProjectVal $
+            p { pMaintainers = Map.delete pk (pMaintainers p) }
+        _ -> v
+
+applyTxPayload TxCheckpoint{} _ _ws = notImplemented
+applyTxPayload TxUpdateContract{} _ _ws = notImplemented
+
+applyTxPayload (TxTransfer _ _ bal) _ _ | bal == 0 =
+    Left (ErrInvalidTransfer bal)
+applyTxPayload (TxTransfer sender receiver bal) author ws = do
+    senderAcc   <- lookupAccount sender   ws
+    receiverAcc <- lookupAccount receiver ws
+
+    when (author /= sender) $
+        Left (ErrNotAuthorized author)
+
+    when (bal > accountBalance senderAcc) $
+        Left (ErrInsufficientBalance (accountBalance senderAcc))
+
+    when (bal > maxBound - accountBalance receiverAcc) $
+        Left (ErrOverflow receiver)
+
+    pure (transfer, TxOutput)
+  where
+    senderKey   = addressKey sender
+    receiverKey = addressKey receiver
+
+    transfer =
+        alter (withAccount (withdraw bal)) senderKey   .
+        alter (withAccount (deposit bal))  receiverKey $ ws
+
+    deposit  n acc = acc { accountBalance = accountBalance acc + n }
+    withdraw n acc = acc { accountBalance = accountBalance acc - n }
+
+    withAccount f v = case v of
+        -- Account exists. Update balance or delete account if zero balance.
+        Just (AccountVal acc) -> case f acc of
+            acc'@Account{..} | accountBalance == 0 -> Nothing
+                             | otherwise           -> Just (AccountVal acc')
+        -- Account doesn't exist. Create it.
+        Nothing -> Just . AccountVal $ (mkAccount receiver) { accountBalance = bal }
+
+        -- Not an account ...
+        val -> val
```

Comment:
This should probably be an error

Reply:
I'm going to return `undefined` here, as this branch should never be reached (we lookup the accounts prior) and that's the idiomatic way to signal that with Protolude. In the next PR we can look into returning an Either from this function, and whether that will make things cleaner.